### PR TITLE
Fix menu anchor state handling

### LINKED ISSUE #157606: `MenuAnchor.builder` does not update properly
The MenuAnchor's [child](https://main-api.flutter.dev/flutter/material/MenuAnchor/child.html) property is documented as follows:

> Supply this child if there is a portion of the widget tree built in [builder](https://main-api.flutter.dev/flutter/material/MenuAnchor/builder.html) that doesn't depend on the controller or context supplied to the [builder](https://main-api.flutter.dev/flutter/material/MenuAnchor/builder.html). It will be more efficient, since Flutter doesn't then need to rebuild this child when those change.

<br>

### Expected Behavior

The builder updates each time the controller opens or closes.

### Actual Behavior

The builder does not update when the controller opens, but does update when it closes.

<br>

### Code Sample

DartPad link: https://dartpad.dev/?id=a0e913ed86a5cf11e223abda4ebd4535&channel=main

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    const MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: MenuAnchor(
            menuChildren: [
              MenuItemButton(child: Text('menu item')),
            ],
            builder: anchorBuilder,
          ),
        ),
      ),
    ),
  );
}

Widget anchorBuilder(
  BuildContext context,
  MenuController controller,
  Widget? child,
) {
  final bool isOpen = controller.isOpen;
  print(isOpen ? 'menu open!' : 'menu closed!');

  return FilledButton(
    onPressed: () {
      // Using controller.isOpen is necessary here, since
      // the value of isOpen does not update properly.
      controller.isOpen ? controller.close() : controller.open();
    },
    child: Text(isOpen ? 'close menu' : 'open menu'),
  );
}
```

### PR DESCRIPTION
**Pull request by YeungKC**
This commit refactors the `_MenuAnchorState` class in `menu_anchor.dart` to include a check for the mounted state and the scheduler phase before calling `setState()`. This ensures that UI updates are only performed when the widget is still mounted and not during the persistent callbacks phase. Additionally, a new test case is added in `menu_anchor_test.dart` to verify that the `isOpen` state of the `MenuAnchor` widget is updated correctly when the button is pressed.

Fix: #157606

*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue. An issue is not required if the PR fixes something trivial like a typo.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/material/menu_anchor.dart [hunk 1]
```diff
@@ -576,6 +576,12 @@ class _MenuAnchorState extends State<MenuAnchor> {
     }
 
     widget.onOpen?.call();
+
+    if (mounted && SchedulerBinding.instance.schedulerPhase != SchedulerPhase.persistentCallbacks) {
+      setState(() {
+        // Mark dirty to ensure UI updates
+      });
+    }
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/menu_anchor.dart:584]: I had no idea that the `SchedulerPhase` could be used this way, thanks!

File: packages/flutter/lib/src/material/menu_anchor.dart [hunk 2]
```diff
@@ -915,7 +915,8 @@ class MenuItemButton extends StatefulWidget {
   /// An optional Semantics label, applied to the entire [MenuItemButton].
   ///
   /// A screen reader will default to reading the derived text on the
-  /// [MenuItemButton] itself, which is not guaranteed to be readable. (For some shortcuts, such as comma, semicolon, and other
+  /// [MenuItemButton] itself, which is not guaranteed to be readable.
+  /// (For some shortcuts, such as comma, semicolon, and other
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/menu_anchor.dart:919]: (Was in the process of suggesting this lol)

**Review by nate-thegrate — APPROVED** [review]
Thanks for the fix, this looks awesome!

**Comment by nate-thegrate** [comment]
Actually, quick question before we land this one—there looks to be a `_MenuAnchorScope` class set up to notify when the value of `isOpen` changes.

https://github.com/flutter/flutter/blob/add7e81f4f5cb9903cbfe583f06e08f4566c5639/packages/flutter/lib/src/material/menu_anchor.dart#L2533-L2551

Currently it doesn't look like the inherited widget rebuilds when the menu is opened; maybe we should remove the `isOpen` field to prevent confusion.

**Comment by YeungKC** [comment]
The _MenuAnchorScope is indeed designed to notify its dependents when the isOpen state changes, but this notification doesn't directly trigger a rebuild of the builder. Instead, it triggers the didChangeDependencies method in the dependent widgets.

**Comment by nate-thegrate** [comment]
> The _MenuAnchorScope is indeed designed to notify its dependents when the isOpen state changes, but this notification doesn't directly trigger a rebuild of the builder. Instead, it triggers the didChangeDependencies method in the dependent widgets.

True that!

I think what I'm trying to get to is: this bug indicates that `_MenuAnchorState` does not rebuild when the menu is opened.

Perhaps it'd be good to remove the `_MenuAnchorScope.isOpen` flag, since currently it isn't set up to report accurate values. Alternatively, we could rework `_MenuAnchorScope` so it's rebuilt when the controller opens/closes.

**Comment by YeungKC** [comment]
Consider a scenario where we're not using the builder pattern, but instead using a simple child widget. After this PR is merged, such a child widget would still be able to detect updates through didChangeDependencies.

**Comment by nate-thegrate** [comment]
> Consider a scenario where we're not using the builder pattern, but instead using a simple child widget. After this PR is merged, such a child widget would still be able to detect updates through didChangeDependencies.

Perhaps so, though we'd need some way to let the child depend on `_MenuAnchorScope` through a public API.

<br>

My concern is:
```dart
context.dependOnInheritedWidgetOfExactType<_MenuAnchorScope>()!.isOpen;
```
In order for this value to be accurate, the inherited widget needs to be rebuilt when the controller opens/closes. And since rebuilding the `_MenuAnchorState` would also involve re-evaluating the `builder`, I believe this bug is evidence that `_MenuAnchorScope.isOpen` is sometimes inaccurate.

Since it's a private class and `isOpen` isn't accessed anywhere, perhaps we should remove the field.

**Comment by YeungKC** [comment]
I don't believe that `_MenuAnchorScope.isOpen` would be inaccurate. Could you provide a test case demonstrating this?

While I understand that `_MenuAnchorScope.isOpen` can't be accessed directly, triggering `didChangeDependencies` seems more "correct" to me. If I were to use `didChangeDependencies` as a trigger and then check the controller's isOpen state, that would also be a valid approach.

I think we may be straying too far from the original intent of this PR. Perhaps it would be better to open a separate issue to discuss this particular concern in more depth.

**Comment by nate-thegrate** [comment]
> I think we may be straying too far from the original intent of this PR.

Fair point, especially if we end up removing the `_MenuAnchorScope.isOpen` field.

If we decide to keep it, however, updating `_MenuAnchorState` to rebuild when its value changes would mean that the changes in this PR are unnecessary.

**Comment by YeungKC** [comment]
I see your point. However, I think there's an important aspect to consider here. Without this PR, the `_MenuAnchorScope` wouldn't update its isOpen state, and consequently, `updateShouldNotify` wouldn't return true when the menu opens or closes.

**Comment by nate-thegrate** [comment]
> I see your point. However, I think there's an important aspect to consider here. Without this PR, the `_MenuAnchorScope` wouldn't update its isOpen state, and consequently, `updateShouldNotify` wouldn't return true when the menu opens or closes.

100% agree. In my mind I was thinking "if `_MenuAnchorScope` is updated to rebuild when the controller opens/closes, then we could just add a dependency to the `builder`"

What didn't connect for me till just now: since this change calls `_MenuAnchorState.setState`, the inherited widget is now guaranteed to be up-to-date, so that's pretty awesome 👍

**Review by bleroux — APPROVED** [review]
LGTM! Thanks for the contribution.

**Pull request merged**